srml/scored-pool/src/lib.rs

```diff
@@ -0,0 +1,368 @@
+		for enum Call
+		where origin: T::Origin
+	{
+		fn deposit_event<T, I>() = default;
+
+		/// Every `Period` blocks the `Members` set is refreshed from the
+		/// highest scoring members in the pool.
+		fn on_initialize(n: T::BlockNumber) {
+			if n % T::Period::get() == Zero::zero() {
+				<Module<T, I>>::refresh_members(true);
+			}
+		}
+
+		/// Add `origin` to the pool of candidates.
+		pub fn issue_candidacy(origin) {
+			let who = ensure_signed(origin)?;
+
+			let _ = Self::find_in_pool(&who)
+				.ok()
+				.map_or_else(
+					|| Ok(()),
+					|_| Err("already a member"),
+				)?;
+
+			T::Currency::reserve(&who, T::CandidateDeposit::get())
+				.map_err(|_| "balance too low")?;
+
+			let mut pool = <Pool<T, I>>::get();
+			pool.push((who.clone(), None));
+			Self::sort_pool(&mut pool);
+			<Pool<T, I>>::put(&pool);
+
+			Self::deposit_event(RawEvent::CandidateAdded);
+		}
+
+		/// An entity withdraws candidacy and gets its deposit back.
+		///
+		/// If the entity is part of the `Members`, then the highest member
+		/// of the `Pool` that is not currently in `Members` is immediately
+		/// placed in the set instead.
+		pub fn withdraw_candidacy(origin) {
+			let who = ensure_signed(origin)?;
+
+			Self::remove_member(who)?;
+			Self::deposit_event(RawEvent::CandidateWithdrew);
+		}
+
+		/// Kick a member `who` from the set.
+		///
+		/// May only be called from `KickOrigin` or root.
+		pub fn kick(origin, who: T::AccountId) {
+			T::KickOrigin::try_origin(origin)
+				.map(|_| ())
+				.or_else(ensure_root)
+				.map_err(|_| "bad origin")?;
+
+			Self::remove_member(who)?;
+			Self::deposit_event(RawEvent::CandidateKicked);
+		}
+
+		/// Score a member `who` with `score`.
+		///
+		/// May only be called from `ScoreOrigin` or root.
+		pub fn score(origin, who: T::AccountId, score: T::Score) {
+			T::ScoreOrigin::try_origin(origin)
+				.map(|_| ())
+				.or_else(ensure_root)
+				.map_err(|_| "bad origin")?;
+
+			let mut pool = <Pool<T, I>>::get();
+			let location = Self::find_in_pool(&who)?;
+			pool.remove(location);
+			pool.push((who.clone(), Some(score.clone())));
+			Self::sort_pool(&mut pool);
+
+			<Pool<T, I>>::put(&pool);
+		}
+
+		/// Root-dispatchable call to change `MemberCount`.
+		pub fn change_member_count(origin, count: u32) {
+			ensure_root(origin)?;
+			<MemberCount<I>>::put(&count);
+		}
+	}
+}
+
+impl<T: Trait<I>, I: Instance> Module<T, I> {
+
+	/// Fetches the `MemberCount` highest scoring members from
+	/// `Pool` and puts them into `Members`.
+	///
+	/// If `notify` is set to true `T::MembershipChanged::change_members_sorted`
+	/// will be invoked at the end of the method.
+	fn refresh_members(notify: bool) {
+		let count = <MemberCount<I>>::get();
+
+		let pool = <Pool<T, I>>::get();
+		let mut new_members: Vec<T::AccountId> = pool
+			.into_iter()
+			.filter(|(_, score)| score.is_some())
+			.rev()
+			.take(count as usize)
+			.map(|(account_id, _)| account_id)
+			.collect();
+		new_members.sort();
```

Comment:
hmmm can't we prove that it already is since it is the`.rev().take(x)` of sth which was already sorted?

Reply:
Unfortunately not because of this:

`Pool = Vec(T::AccountId, Option<T::Score>>` is sorted by score, not by account id.
`new_members` is only `Vec<T::AccountId>` and needs to be sorted by account_id since `ChangeMembers::change_members_sorted()` requires the input to be sorted.

Also the `binary_search` functions which follow in the next lines (used to derive incoming and outgoing elements) require a sorted vec.